jaxrs-spec/src/main/asciidoc/chapters/resources/_declaring_method_capabilities.adoc

```diff
@@ -126,3 +129,65 @@ the media type `application/xml` given its higher qs-value. Note that
 qs-values, just like q-values, are relative and as such are only
 comparable to other qs-values within the same `@Produces` annotation
 instance. For more information see <<determine_response_type>>.
+
+[[consuming_multipart_formdata]]
+==== Special handling for consuming multipart/form-data media types
+
+The `multipart/form-data` media type enables requests to send multiple
+entities (parts) as a single entity. Each part contains its own set of
+headers, media type and content. There are a few different ways to
+consume these requests in a resource method. First, a resource method
+may contain an entity parameter of type, `List<Part>`. For example:
+
+[source,java]
+----
+@Path("widgets3")
+public class WidgetsResource3 {
+
+    @POST
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public Response postWidget(List<Part> parts) {
+        for (Part part : parts) {
+            String name = part.getName();
+            Optional<String> fileName = part.getFileName();
+            InputStream is = part.getContent();
+            MultivaluedMap<String, String> partHeaders = part.getHeaders();
+            MediaType mediaType = part.getMediaType();
+            doSomethingWithPart(name, fileName, is, partHeaders, mediaType);
+        }
+        return Response.ok().build();
+    }
+
+}
+----
+
+Another approach is to use `@FormParam` parameters where the value in the
+annotation corresponds to the name of the part. The parameter type may be
+a `jakarta.ws.rs.ext.Part`, a `java.io.InputStream`, a `String`, or any type
+that can be converted to using a registered `MessageBodyReader` or
+`ParamConverter`. Here is
```

Comment:
Sorry. One last question. 

The API docs of `Part#getContent(Class|GenericType)` only mentions `MessageBodyReader`, but here we allow `ParamConverter` as well.

I'm wondering if `ParamConverter` makes sense here. If we allow both, we will most likely have to define which one of the two mechanisms is preferred over the other. And `ParamConverter` uses a `String` as the input and in the next paragraph we explicitly warn about `String` being not optimal as it requires to read the full entity into the heap.

Reply:
My thinking here was that `ParamConverter` providers are already responsible for converting Strings to parameter types.  The only real difference here (with `@FormParam`) is that we have an `InputStream` instead of a a `String` - but since we allow the implementation to convert the `InputStream` to a `String` for `@FormParam(...) String`, it seemed intuitive (to me, at least :) ) that the implementation should first check that it can convert to the parameter type using registered MBRs, and if it finds none, then it could check to see if a `ParamConverter` is registered for the parameter type.  If so, then the implementation would convert the part's `InputStream` to `String` and then invoke the `ParamConverter`.  

For `Part#getContent(...)`, I don't think it makes sense to use `ParamConverter` because at this point, it's not a parameter; the parameter was `Part`.  A MBR does make sense because it is reading the message body of the individual part to the specified class or `GenericType`.

Reply:
BTW, I appreciate the thoroughness of your review.  I'd certainly rather spend time up front to get it right rather than trying to patch something that was broken from the start.  :)

Reply:
@andymc12 Thanks for your explanation and sorry for the delayed reply.

After thinking about this more, it somehow makes sense to allow `ParamConverter` in case of `@FormParam`, as it is also used for standard `@FormParam` cases. However, I also think it could be a bit weird for the user that `@FormParam` and `getContent` may behave differently.

So basically we have three options:

  1. `@FormParam` uses the MBR and falls back to a `ParamConverter` (if available), but `getContent()` only uses the MBR.
  2. Both `@FormParam` and `getContent()` use the MBR and fall back to `ParamConverter`.
  3. Both `@FormParam` and `getContent()` only use the MBR.

To be honest, I'm not sure yet which option I personally like most.

@eclipse-ee4j/ee4j-jaxrs-committers Other opinions?

Reply:
Documentation of `ParamConverter` says it applies to parameters injected via (among other annotations) `@FormParam`.:

> Defines a contract for a delegate responsible for converting between a String form of a message parameter value and the corresponding custom Java type T. Conversion of message parameter values injected via `@PathParam`, `@QueryParam`, `@MatrixParam`, `@FormParam`, `@CookieParam` and `@HeaderParam` is supported. 

I am not aware of an existing API method that would utilize converters. Therefore to maintain consistency, I'd suggest staying with option 1.

Reply:
> 1. `@FormParam` uses the MBR and falls back to a `ParamConverter` (if available), but `getContent()` only uses the MBR.
> 2. Both `@FormParam` and `getContent()` use the MBR and fall back to `ParamConverter`.
> 3. Both `@FormParam` and `getContent()` only use the MBR.

`getContent()` is not intended for parameters, so it shouldn't use `ParamConverter`.

My preference would be 1.

Actually after some thinking I wonder why we want `@FormParam` to use MBR, actually, as it is a parameter.

Reply:
In my view `@FormParam` on types other than `EntityPart` is basically a convenient way to parse and access part data.

So instead of doing this:

```java
public Response someMethod( List<EntityPart> parts ) {

  JsonObject data = parts.stream()
    .filter( p -> p.getId().equals("file") )
    .findFirst()
    .map( p -> p.getContent( JsonObject.class ) )
    .orElse( null );

   // ...

}
```

or instead of this:

```java
public Response someMethod( @FormParam("file") EntityPart part ) {

  JsonObject data = part.getContent( JsonObject.class );
  // ...

}
```


You can simply do this:

```java
public Response someMethod( @FormParam("file") JsonObject data ) {
  // ...
}
```

And therefore it would make sense that using `@FormParam` on some specific type `T` is the same as calling `EntityPart.getContent(T.class)`.

However, I agree that this may be problematic given the originally intended usage of `ParamConverter` and `MessageBodyReader`.